Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2021, timmot <tiwwot@protonmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Debug.h>
+#include <FileSystemAccessServer/ClientConnection.h>
+#include <LibCore/File.h>
+#include <LibCore/IODevice.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/FilePicker.h>
+#include <LibGUI/MessageBox.h>
+
+namespace FileSystemAccessServer {
+
+static HashMap<int, NonnullRefPtr<ClientConnection>> s_connections;
+
+ClientConnection::ClientConnection(NonnullRefPtr<Core::LocalSocket> socket, int client_id)
+    : IPC::ClientConnection<FileSystemAccessClientEndpoint, FileSystemAccessServerEndpoint>(*this, move(socket), client_id)
+{
+    s_connections.set(client_id, *this);
+}
+
+ClientConnection::~ClientConnection()
+{
+}
+
+void ClientConnection::die()
+{
+    s_connections.remove(client_id());
+    GUI::Application::the()->quit();
+    exit(0);
+}
+
+Messages::FileSystemAccessServer::RequestFileResponse ClientConnection::request_file(String const& path, Core::OpenMode const& requested_access)
+{
+    // Must receive an absolute path
+    VERIFY(path.starts_with("/"sv));
+
+    bool approved = false;
+    auto maybe_permissions = m_approved_files.get(path);
+
+    if (maybe_permissions.has_value())
+        approved = has_flag(maybe_permissions.value(), requested_access);
+
+    if (!approved) {
+        StringBuilder builder;
+        if (has_flag(requested_access, Core::OpenMode::ReadOnly))
+            builder.append('r');
+
+        if (has_flag(requested_access, Core::OpenMode::WriteOnly))
+            builder.append('w');
+
+        auto access_string = builder.to_string();
+
+        auto pid = this->socket().peer_pid();
+        auto exe_link = LexicalPath("/proc").append(String::number(pid)).append("exe").string();
+        auto exe_path = Core::File::real_path_for(exe_link);
+        auto exe_name = LexicalPath(exe_path).basename();
```

Comment:
Since `basename()` returns a `StringView const&`, the memory pointed to by `exe_name` may already be freed. You should use `LexicalPath::basename(exe_path)`, which returns `String` instead of `StringView const&`.